runtime/gc_base/IndexableObjectAllocationModel.cpp

```diff
@@ -264,17 +264,28 @@ MM_IndexableObjectAllocationModel::layoutDiscontiguousArraylet(MM_EnvironmentBas
 	if (NULL != spine) {
 		switch (_layout) {
 		case GC_ArrayletObjectModel::Discontiguous:
-			/* if last arraylet leaf is empty (contains 0 bytes) arrayoid pointer is set to NULL */
-			if (arrayoidIndex == (_numberOfArraylets - 1)) {
-				Assert_MM_true(0 == (_dataSize % arrayletLeafSize));
-				GC_SlotObject slotObject(env->getOmrVM(), &(arrayoidPtr[arrayoidIndex]));
-				slotObject.writeReferenceToSlot(NULL);
-			} else {
-				Assert_MM_true(0 != (_dataSize % arrayletLeafSize));
+#if defined(LINUX)
+#if defined(J9VM_GC_ENABLE_DOUBLE_MAP)
+			/* All data from array are stored in the leaves */
+			if(extensions->indexableObjectModel.isDoubleMappingEnabled()) {
 				Assert_MM_true(arrayoidIndex == _numberOfArraylets);
+			} else
+#endif /* J9VM_GC_ENABLE_DOUBLE_MAP */
+#endif /* LINUX */
+			{
+				 /* if last arraylet leaf is empty (contains 0 bytes) arrayoid pointer is set to NULL */
+				if (arrayoidIndex == (_numberOfArraylets - 1)) { 
+					Assert_MM_true(0 == (_dataSize % arrayletLeafSize));
+					GC_SlotObject slotObject(env->getOmrVM(), &(arrayoidPtr[arrayoidIndex]));
+					slotObject.writeReferenceToSlot(NULL);
+				} else { 
+					Assert_MM_true(0 != (_dataSize % arrayletLeafSize));
+					Assert_MM_true(arrayoidIndex == _numberOfArraylets);
+				}
 			}
 			break;
 
+		/* Unreachable if double map is enabled */
```

Comment:
Can this be turned into a trace assert?

Reply:
If I understand correctly and this comment relates to `case GC_ArrayletObjectModel::Hybrid:` it should be moved two lines lower after `#if defined(J9VM_GC_ENABLE_DOUBLE_MAP)`